### PR TITLE
n-Rooks sampling

### PR DESCRIPTION
**Pull request by simonkrogmann**
This includes 2D n-Rooks sampling  and a compile test.

### CODE REVIEW

**Comment by scheibel** [comment]
:shipit:

File: source/glkernel/include/glkernel/sample.h [hunk 1]
```diff
@@ -22,6 +22,9 @@ template <typename T, glm::precision P>
 size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, T min_dist, unsigned int num_probes = 32);
 
 
+template <typename T, glm::precision P>
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.h]: please add comment in the following style:

``` c++
    /**
    *  @brief
    *    Bla ...
    *
    *  @param[in] name
    *    Bla ...
    *
    *  @return
    *    Bla ...
    */
```

in addition, comment on algorithm

* generate one jittered sample per row and column
* randomly shuffle the samples in x- or y-direction
* uniform distribution of the projections in x- and y-direction
* can generate an arbitrary number of stratified samples

(taken from http://cg.informatik.uni-freiburg.de/course_notes/graphics2_04_sampling.pdf)

**Comment by cgcostume** [comment]
please adhere to [#2.2](http://cginternals.github.io/guidelines/cpp/#2.2)

File: source/glkernel/include/glkernel/sample.hpp [hunk 2]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
     return k + 1;
 }
 
+template <typename T, glm::precision P>
+size_t n_rooks(tkernel<glm::tvec2<T, P>> & kernel)
+{
+	assert(kernel.depth() == 1);
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: same should be done in #1 - perhaps we can extend the signature to allow for an arbitrary "layer" at depth?

File: source/glkernel/include/glkernel/sample.hpp [hunk 3]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+	assert(kernel.width() == kernel.height());
+
+	auto stratum_size = 1.0 / kernel.width();
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: please adhere to [7.11](http://cginternals.github.io/guidelines/cpp/#7.11)

File: source/glkernel/include/glkernel/sample.hpp [hunk 4]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+	std::random_device RD;
+	std::mt19937_64 generator(RD());
+	std::uniform_real_distribution<> jitter_dist(0.0, stratum_size);
+
+	std::vector<unsigned int> pool;
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: comment this: e.g., "Pool of row values, column shuffling is done independently, see below"

File: source/glkernel/include/glkernel/sample.hpp [hunk 5]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+	for (unsigned int k = 0; k < kernel.width(); ++k)
+	{
+		pool.push_back(k);
+	}
+	std::random_shuffle(pool.begin(), pool.end());
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: hmm, :smile: perhaps we can use glkernel to do this in the future? we should make a decision on whether or not to use kernels internally for the manipulation of kernels ... in this case this is currently hindered due to missing support of integer coefficients in kernels ...

File: source/glkernel/include/glkernel/sample.hpp [hunk 6]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+
+	for (auto k = 0; k < kernel.width(); ++k)
+	{
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: imo, ```#pragma omp parallel for``` should be applicable here

File: source/glkernel/include/glkernel/sample.hpp [hunk 7]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+		auto x_coord = k * stratum_size + jitter_dist(generator);
+		auto y_coord = pool.at(k) * stratum_size + jitter_dist(generator);
+		auto sample = glm::tvec2<T, P>(x_coord, y_coord);
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: please adhere to [7.11](http://cginternals.github.io/guidelines/cpp/#7.11)

File: source/glkernel/include/glkernel/sample.hpp [hunk 8]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+	assert(kernel.depth() == 1);
+	assert(kernel.width() == kernel.height());
+
+	auto stratum_size = 1.0 / kernel.width();
+	std::random_device RD;
+	std::mt19937_64 generator(RD());
+	std::uniform_real_distribution<> jitter_dist(0.0, stratum_size);
+
+	std::vector<unsigned int> pool;
+	for (unsigned int k = 0; k < kernel.width(); ++k)
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: we need to think about the kernel semantics here: hmm, as far as i understand the n-rooks sampling, the number of samples should be ```kernel.size()```. It should work for any kernel dimension (and works for 2, 3, as well as for 4 coefficients ?, but we can ignore this for now) ... 

I would expect the function to fill all values within my kernel - having a 2D kernel of arbitrary width and height, it should result in width * height samples that use #size strata vertically and #size strata horizontally ... ?

File: source/glkernel/include/glkernel/sample.hpp [hunk 9]
```diff
@@ -207,6 +208,33 @@ size_t poisson_square(tkernel<glm::tvec2<T, P>> & kernel, const T min_dist, cons
...
+		pool.push_back(k);
+	}
+	std::random_shuffle(pool.begin(), pool.end());
+
+	for (auto k = 0; k < kernel.width(); ++k)
+	{
+		auto x_coord = k * stratum_size + jitter_dist(generator);
+		auto y_coord = pool.at(k) * stratum_size + jitter_dist(generator);
+		auto sample = glm::tvec2<T, P>(x_coord, y_coord);
+		kernel[k] = sample;
```
Comment (cgcostume) [inline source/glkernel/include/glkernel/sample.hpp]: prefer to directly assign the sample (reduce Lines-of-Code)

**Comment by cgcostume** [comment]
similar to #1 n rooks sampling can be tested by verifying the resulting kernel, this is also a good way to additionally indicate the intended use of this sampling (as well as be clear about this in the functions declaration commentary).

**Comment by cgcostume** [comment]
test this please

**Comment by cgcostume** [comment]
jenkins, test this please!

**Comment by cgcostume** [comment]
jenkins, test this please!

**Comment by cgcostume** [comment]
jenkins, test this please!

**Comment by cgcostume** [comment]
jenkins, test this please!

**Pull request closed without merging**